nimble/ll: Refactor transitions

MariuszSkamra opened the pull request:
This PR is rebased version of https://github.com/apache/mynewt-nimble/pull/1974 with additional change that introduces Rx early LL ISR callback.

New Bluetooth features require fast RX-to-RX and TX-to-TX transitions and use a different TIFS than 150us.

mkasenberg committed:
nimble/ll: Refactor transitions

New Bluetooth features require fast RX-to-RX and TX-to-TX transitions
and use a different TIFS than 150us.

Co-authored-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

MariuszSkamra committed:
nimble/ll: Introduce Rx end early callback

This adds Rx end early callback that is called before scheduled
PHY transition. This allows user to e.g. prepare CCM for next Rx.

codecoup-tester commented:
Scheduled PR https://github.com/apache/mynewt-nimble/pull/2107#issuecomment-3311800089, board: nrf53, estimated start time: 13:18:19, test case count: 100, estimated duration: 0<details><summary>Test cases to be run</summary>GAP/BROB/BCST/BV-01-C<br>GAP/BROB/BCST/BV-02-C<br>GAP/BROB/BCST/BV-03-C<br>GAP/BROB/BCST/BV-04-C<br>GAP/BROB/BCST/BV-05-C<br>GAP/BROB/OBSV/BV-01-C<br>GAP/BROB/OBSV/BV-02-C<br>GAP/BROB/OBSV/BV-05-C<br>GAP/BROB/OBSV/BV-06-C<br>GAP/DISC/NONM/BV-01-C<br>GAP/DISC/NONM/BV-02-C<br>GAP/DISC/LIMM/BV-03-C<br>GAP/DISC/LIMM/BV-04-C<br>GAP/DISC/GENM/BV-03-C<br>GAP/DISC/GENM/BV-04-C<br>GAP/DISC/LIMP/BV-01-C<br>GAP/DISC/LIMP/BV-02-C<br>GAP/DISC/LIMP/BV-03-C<br>GAP/DISC/LIMP/BV-04-C<br>GAP/DISC/LIMP/BV-05-C<br>GAP/DISC/GENP/BV-01-C<br>GAP/DISC/GENP/BV-02-C<br>GAP/DISC/GENP/BV-03-C<br>GAP/DISC/GENP/BV-04-C<br>GAP/DISC/GENP/BV-05-C<br>GAP/DISC/RPA/BV-01-C<br>GAP/IDLE/NAMP/BV-01-C<br>GAP/IDLE/NAMP/BV-02-C<br>GAP/CONN/NCON/BV-01-C<br>GAP/CONN/NCON/BV-02-C<br>GAP/CONN/NCON/BV-03-C<br>GAP/CONN/DCON/BV-01-C<br>GAP/CONN/DCON/BV-04-C<br>GAP/CONN/DCON/BV-05-C<br>GAP/CONN/UCON/BV-01-C<br>GAP/CONN/UCON/BV-02-C<br>GAP/CONN/UCON/BV-03-C<br>GAP/CONN/UCON/BV-06-C<br>GAP/CONN/ACEP/BV-01-C<br>GAP/CONN/ACEP/BV-03-C<br>GAP/CONN/ACEP/BV-04-C<br>GAP/CONN/GCEP/BV-01-C<br>GAP/CONN/GCEP/BV-02-C<br>GAP/CONN/GCEP/BV-05-C<br>GAP/CONN/GCEP/BV-06-C<br>GAP/CONN/SCEP/BV-01-C<br>GAP/CONN/SCEP/BV-03-C<br>GAP/CONN/DCEP/BV-01-C<br>GAP/CONN/DCEP/BV-03-C<br>GAP/CONN/DCEP/BV-05-C<br>GAP/CONN/DCEP/BV-06-C<br>GAP/CONN/CPUP/BV-01-C<br>GAP/CONN/CPUP/BV-02-C<br>GAP/CONN/CPUP/BV-03-C<br>GAP/CONN/CPUP/BV-04-C<br>GAP/CONN/CPUP/BV-05-C<br>GAP/CONN/CPUP/BV-06-C<br>GAP/CONN/CPUP/BV-08-C<br>GAP/CONN/CPUP/BV-10-C<br>GAP/CONN/TERM/BV-01-C<br>GAP/CONN/PRDA/BV-01-C<br>GAP/CONN/PRDA/BV-02-C<br>GAP/BOND/NBON/BV-01-C<br>GAP/BOND/NBON/BV-02-C<br>GAP/BOND/NBON/BV-03-C<br>GAP/BOND/BON/BV-01-C<br>GAP/BOND/BON/BV-02-C<br>GAP/BOND/BON/BV-03-C<br>GAP/BOND/BON/BV-04-C<br>GAP/SEC/SEM/BV-21-C<br>GAP/SEC/SEM/BV-37-C<br>GAP/SEC/SEM/BV-38-C<br>GAP/SEC/SEM/BV-22-C<br>GAP/SEC/SEM/BV-39-C<br>GAP/SEC/SEM/BV-40-C<br>GAP/SEC/SEM/BV-23-C<br>GAP/SEC/SEM/BV-24-C<br>GAP/SEC/SEM/BV-56-C<br>GAP/SEC/SEM/BV-57-C<br>GAP/SEC/SEM/BV-58-C<br>GAP/SEC/SEM/BV-59-C<br>GAP/SEC/SEM/BV-60-C<br>GAP/SEC/SEM/BV-61-C<br>GAP/SEC/SEM/BV-26-C<br>GAP/SEC/SEM/BV-41-C<br>GAP/SEC/SEM/BV-42-C<br>GAP/SEC/SEM/BV-27-C<br>GAP/SEC/SEM/BV-43-C<br>GAP/SEC/SEM/BV-44-C<br>GAP/SEC/SEM/BV-28-C<br>GAP/SEC/SEM/BV-29-C<br>GAP/SEC/SEM/BV-62-C<br>GAP/SEC/SEM/BV-63-C<br>GAP/SEC/SEM/BV-64-C<br>GAP/SEC/SEM/BV-65-C<br>GAP/SEC/SEM/BV-66-C<br>GAP/SEC/SEM/BV-67-C<br>GAP/SEC/SEM/BI-09-C<br>GAP/SEC/SEM/BI-10-C<br>GAP/SEC/SEM/BI-13-C</details>
<details><summary>Test cases skipped due to limit, count: 59</summary>GAP/SEC/SEM/BV-45-C<br>GAP/SEC/SEM/BI-20-C<br>GAP/SEC/SEM/BI-21-C<br>GAP/SEC/SEM/BI-22-C<br>GAP/SEC/SEM/BI-23-C<br>GAP/SEC/AUT/BV-11-C<br>GAP/SEC/AUT/BV-12-C<br>GAP/SEC/AUT/BV-13-C<br>GAP/SEC/AUT/BV-14-C<br>GAP/SEC/AUT/BV-17-C<br>GAP/SEC/AUT/BV-18-C<br>GAP/SEC/AUT/BV-19-C<br>GAP/SEC/AUT/BV-20-C<br>GAP/SEC/AUT/BV-21-C<br>GAP/SEC/AUT/BV-22-C<br>GAP/SEC/AUT/BV-23-C<br>GAP/SEC/AUT/BV-24-C<br>GAP/SEC/AUT/BV-25-C<br>GAP/SEC/AUT/BV-26-C<br>GAP/SEC/AUT/BV-27-C<br>GAP/SEC/AUT/BV-28-C<br>GAP/PRIV/CONN/BV-10-C<br>GAP/PRIV/CONN/BV-11-C<br>GAP/PRIV/CONN/BV-12-C<br>GAP/ADV/BV-01-C<br>GAP/ADV/BV-02-C<br>GAP/ADV/BV-03-C<br>GAP/ADV/BV-04-C<br>GAP/ADV/BV-05-C<br>GAP/ADV/BV-08-C<br>GAP/ADV/BV-09-C<br>GAP/ADV/BV-10-C<br>GAP/ADV/BV-11-C<br>GAP/ADV/BV-12-C<br>GAP/ADV/BV-13-C<br>GAP/ADV/BV-14-C<br>GAP/ADV/BV-17-C<br>GAP/ADV/BV-18-C<br>GAP/ADV/BV-19-C<br>GAP/PADV/PASM/BV-01-C<br>GAP/PADV/PAM/BV-01-C<br>GAP/PADV/PASE/BV-01-C<br>GAP/PADV/PASE/BV-02-C<br>GAP/PADV/PASE/BV-03-C<br>GAP/PADV/PASE/BV-04-C<br>GAP/PADV/PASE/BV-05-C<br>GAP/PADV/PASE/BV-06-C<br>GAP/PADV/PAST/BV-01-C<br>GAP/PADV/PAST/BV-02-C<br>GAP/GAT/BV-04-C<br>GAP/GAT/BV-05-C<br>GAP/GAT/BV-06-C<br>GAP/GAT/BV-16-C<br>GAP/GAT/BV-17-C<br>GAP/GAT/BV-18-C<br>GAP/BIS/BSE/BV-01-C<br>GAP/BIS/BBM/BV-01-C<br>GAP/CSUB/CSR/BV-01-C<br>GAP/CSUB/CSU/BV-01-C</details>

MariuszSkamra committed:
nimble/ll: Factor out PHY transitions to helper functions

This moves transition-related code to dedicated Rx/Tx functions to reduce
function size and remove duplicated code. No functional change intended.

Co-authored-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Co-authored-by: Magdalena Kasenberg <magdalena.kasenberg@codecoup.pl>

MariuszSkamra committed:
nimble/ll: Remove unused BLE_PHY_TRANSITION_RX_TX definition

This removes unused BLE_PHY_TRANSITION_RX_TX definition from API.

MariuszSkamra committed:
nimble/ll: Rename PHY transitions

This renames transition definitions:
BLE_PHY_TRANSITION_TX_RX -> BLE_PHY_TRANSITION_TO_RX
BLE_PHY_TRANSITION_TX_TX -> BLE_PHY_TRANSITION_TO_TX

MariuszSkamra commented:
@andrzej-kaczmarek
